pkg/providers/pricing/pricing.go

```diff
@@ -225,85 +227,87 @@ func (p *DefaultProvider) UpdateOnDemandPricing(ctx context.Context) error {
+			Type:  "TERM_MATCH",
 			Value: aws.String("OnDemand"),
 		}},
 		additionalFilters...)
 
-	err := p.pricing.GetProductsPagesWithContext(
-		ctx,
-		&pricing.GetProductsInput{
-			Filters:     filters,
-			ServiceCode: aws.String("AmazonEC2"),
-		},
-		p.onDemandPage(ctx, prices),
-	)
-	if err != nil {
-		return nil, err
+	input := &pricing.GetProductsInput{
+		Filters:     filters,
+		ServiceCode: aws.String("AmazonEC2"),
+	}
+
+	paginator := pricing.NewGetProductsPaginator(p.pricing, input)
+	for paginator.HasMorePages() {
+		output, err := paginator.NextPage(ctx)
+
+		if err != nil {
+			return nil, fmt.Errorf("getting pricing data, %w", err)
+		}
+		prices = lo.Assign(prices, p.onDemandPage(ctx, output))
 	}
 
 	return prices, nil
 }
 
-func (p *DefaultProvider) spotPage(ctx context.Context, prices map[string]map[string]float64) func(output *ec2.DescribeSpotPriceHistoryOutput, b bool) bool {
-	return func(output *ec2.DescribeSpotPriceHistoryOutput, b bool) bool {
-		for _, sph := range output.SpotPriceHistory {
-			spotPriceStr := aws.StringValue(sph.SpotPrice)
-			spotPrice, err := strconv.ParseFloat(spotPriceStr, 64)
-			// these errors shouldn't occur, but if pricing API does have an error, we ignore the record
-			if err != nil {
-				log.FromContext(ctx).V(1).Info(fmt.Sprintf("unable to parse price record %#v", sph))
-				continue
-			}
-			if sph.Timestamp == nil {
-				continue
-			}
-			instanceType := aws.StringValue(sph.InstanceType)
-			az := aws.StringValue(sph.AvailabilityZone)
-			_, ok := prices[instanceType]
-			if !ok {
-				prices[instanceType] = map[string]float64{}
-			}
-			prices[instanceType][az] = spotPrice
+func (p *DefaultProvider) spotPage(ctx context.Context, output *ec2.DescribeSpotPriceHistoryOutput, prices map[string]map[string]float64) map[string]map[string]float64 {
+	for _, sph := range output.SpotPriceHistory {
+		spotPriceStr := aws.ToString(sph.SpotPrice)
+		spotPrice, err := strconv.ParseFloat(spotPriceStr, 64)
+		// these errors shouldn't occur, but if pricing API does have an error, we ignore the record
+		if err != nil {
+			log.FromContext(ctx).V(1).Info(fmt.Sprintf("unable to parse price record %#v", sph))
+			continue
 		}
-		return true
+		if sph.Timestamp == nil {
+			continue
+		}
+		instanceType := string(sph.InstanceType)
+		az := aws.ToString(sph.AvailabilityZone)
+		_, ok := prices[instanceType]
+		if !ok {
+			prices[instanceType] = map[string]float64{}
+		}
+		prices[instanceType][az] = spotPrice
+
 	}
+	return prices
 }
 
 // turning off cyclo here, it measures as a 12 due to all of the type checks of the pricing data which returns a deeply
 // nested map[string]interface{}
 // nolint: gocyclo
-func (p *DefaultProvider) onDemandPage(ctx context.Context, prices map[string]float64) func(output *pricing.GetProductsOutput, b bool) bool {
+func (p *DefaultProvider) onDemandPage(ctx context.Context, output *pricing.GetProductsOutput) (result map[string]float64) {
```

Comment:
```suggestion
func (p *DefaultProvider) onDemandPage(ctx context.Context, output *pricing.GetProductsOutput) map[string]float64 {
```

Do we need to name the return argument?